File: packages/asdl-slots/src/asdl_slots/cli/slot/list.py
from __future__ import annotations

from typing import Literal

import click

from asdl_core import get_console, make_table
from asdl_core.clinkr.ensure import Ensure
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_slots.cli.slot.context import load_slots_context
from asdl_slots.inventory import SlotInventory, build_slot_inventory

SlotStatus = Literal["assigned", "available"]


class SlotListRequest(ClinkrModel):
    """No inputs — `slot list` always renders the whole pool for the repo."""


class SlotRow(ClinkrModel):
    slot_name: str
    branch: str | None
    operation: str | None
    worktree_path: str
    status: SlotStatus


class SlotListResult(ClinkrModel):
    pool_size: int
    rows: tuple[SlotRow, ...]
    repo_name: str


def render_slot_list(result: SlotListResult) -> None:
    table = make_table()
    table.add_column("Slot", style="bold cyan", no_wrap=True)
    table.add_column("Status", no_wrap=True)
    table.add_column("Branch", no_wrap=True, overflow="ellipsis", ratio=1)
    table.add_column("Operation", no_wrap=True)
    table.add_column("Worktree", no_wrap=True, style="dim", overflow="ellipsis", ratio=1)

    for row in result.rows:
        operation = "" if row.operation is None else f"{row.operation} in progress"
        table.add_row(
            row.slot_name,
            row.status,
            row.branch or "",
            operation,
            row.worktree_path,
        )

    get_console().print(table)


def _compose_rows(inventory: SlotInventory) -> tuple[SlotRow, ...]:
    return tuple(
        SlotRow(
            slot_name=record.slot_name,
            branch=record.branch,
            operation=record.operation,
            worktree_path=str(record.path),
            status=record.status,
        )
        for record in inventory.records
    )


@clinkr_operation(
    name="list",
    help="List worktree pool slots derived from Git worktree state.",
    aliases=("ls",),
    human_renderer=render_slot_list,
)
def run_list_slots(ctx: click.Context, request: SlotListRequest) -> ClinkrExit[SlotListResult]:
    slots_ctx = Ensure.ideal_state(load_slots_context(ctx))

    inventory = build_slot_inventory(slots_ctx.git)
    return ClinkrExit.ok(
        SlotListResult(
            pool_size=inventory.pool_size,
            rows=_compose_rows(inventory),
            repo_name=slots_ctx.repo.repo_name,
        )
    )

File: ts/packages/slot/src/operations/list.ts
import { failure, ok, type RenderCapabilities } from "@asdl/clinkr";
import { renderTextTable } from "@asdl/core/text-table";
import { z } from "zod";

import type { SlotCliContext } from "../context.ts";
import { buildSlotInventory, poolSize, slotStatus } from "../inventory.ts";

export const listRequestSchema = z.object({});

export const slotRowSchema = z.object({
	slot_name: z.string(),
	branch: z.string().nullable(),
	operation: z.string().nullable(),
	worktree_path: z.string(),
	status: z.union([z.literal("assigned"), z.literal("available")]),
});

export const listResultSchema = z.object({
	pool_size: z.number().int().nonnegative(),
	rows: z.array(slotRowSchema),
	repo_name: z.string(),
});

export type ListRequest = z.infer<typeof listRequestSchema>;
export type ListResult = z.infer<typeof listResultSchema>;

export async function runList(ctx: SlotCliContext, _request: ListRequest) {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	return ok({
		pool_size: poolSize(inventory),
		repo_name: ctx.repo.repoName,
		rows: inventory.records.map((record) => ({
			slot_name: record.slotName,
			branch: record.branch,
			operation: record.operation,
			worktree_path: record.path,
			status: slotStatus(record),
		})),
	});
}

export function renderList(result: ListResult, caps: RenderCapabilities = { color: false }): string {
	if (result.rows.length === 0) return `No slots initialized for ${result.repo_name}.`;
	return renderTextTable({
		columns: [
			{ header: "SLOT", style: "bold-cyan" },
			{ header: "STATUS" },
			{ header: "BRANCH" },
			{ header: "OPERATION" },
			{ header: "WORKTREE", style: "dim" },
		],
		rows: result.rows.map((row) => [row.slot_name, row.status, row.branch ?? "—", row.operation === null ? "—" : `${row.operation} in progress`, row.worktree_path]),
		color: caps.color,
		rule: true,
		headerStyle: "bold-cyan",
	});
}
